docs/USER_GUIDE.md

```diff
@@ -0,0 +1,506 @@
+```
+
+# Generate API bindings
+
+At first, you need VPP JSON API bindings to build go bindings from.
+Build the VPP locally, or download
+from `packagecloud.io` ([learn more](https://fd.io/docs/vpp/master/gettingstarted/installing)).
+
+JSON API files are located in `/usr/share/vpp/api/`
+
+```
+$ make generate-binapi 
+# installing binapi-generator v0.8.0-alpha-6-g2054a76-dirty
+# generating binapi
+binapi-generator -input-dir=/usr/share/vpp/api -output-dir=. -gen=rpc
+INFO[0000] resolved VPP version from installed package: 23.06-rc0~65-g2ddb2fdaa
+INFO[0000] Generating 236 files                         
+binapi-generator -input-file=/usr/share/vpp/api/core/vpe.api.json -output-dir=. -gen=http
+INFO[0000] resolved VPP version from installed package: 23.06-rc0~65-g2ddb2fdaa
+INFO[0000] Generating 3 files   
+```
+
+Generated files can be found in `/binapi` package.
+
+### Plugins
+
+The binary API generator supports custom plugins generating additional files.
+
+Available plugins:
+
+- `http` generates HTTP handlers
+- `rpc` generates RPC services
+
+### Options
+
+- `version` prints the generator version, for example, `govpp v0.8.0-dev`
+- `gen` allows injecting generator plugin (`rpc` is the default)
+- `import-prefix` generates files with custom import prefix
+- `input-dir` sets the custom input directory instead of the default
+- `output-dir` sets the custom output directory
+- `debug` prints some additional logs
+
+## VPP API calls
+
+GoVPP uses adapters to talk to VPP sockets.
+
+* **socketclient** connects to the VPP `api.sock` and supports all VPP binary APIs
+* **statsclient** connects to the VPP `stats.sock` and reads VPP stats
+
+### Connection
+
+Two connection types to the binary API socket exist - synchronous and asynchronous. GoVpp can connect to more VPPs by
+creating multiple connections.
+
+#### Synchronous connect
+
+The synchronous connect creates a new adapter instance. The call blocks until the connection is established.
+
+```
+conn, err := govpp.Connect(socketPath)
+if err != nil {
+    // handle the error
+}
+defer conn.Disconnect()
+```
+
+#### Asynchronous connect
+
+The main difference between the synchronous and asynchronous connection is that the asynchronous connection does not
+block until the connection is ready. Instead, the caller receives a channel to watch connection events.
+
+```
+conn, connEv, err := govpp.AsyncConnect(socketPath, attemptNum, interval)
+if err != nil {
+    // handle error
+}
+defer conn.Disconnect()
+
+// wait for the connection event
+e := <-connEv
+if e.State != core.Connected {
+   // handle error in e.Error
+}
+```
+
+You might notice additional parameters `attemptNum` and `interval` defining the number of reconnect attempts, and an
+interval (in seconds) between those attempts.
+
+### Sending API messages
+
+Each binary API message in the Go-generated API is a data structure. The caller can send API messages either using
+the `Channel`, or the `Stream client`.
```

Comment:
```suggestion
the `Channel` (legacy method), or the `Stream` (preferred).
```

This should also include links to the discussion #43 that contains more details.